ts/nni_manager/training_service/reusable/trialDispatcher.ts

```diff
@@ -279,64 +296,113 @@ class TrialDispatcher implements TrainingService {
             const environment = environments[index];
             if (environment.isAlive === true) {
                 this.log.info(`stopping environment ${environment.id}...`);
-                await environmentService.stopEnvironment(environment);
-                await this.commandChannel.close(environment);
+                if (environment.environmentService === undefined) {
+                    throw new Error(`${environment.id} do not have environmentService!`);
+                }
+                await environment.environmentService.stopEnvironment(environment);
+                const commandChannel: CommandChannel | undefined = 
+                    this.commandChannelDict.get(environment.environmentService.getCommandChannelName);
+                if (commandChannel === undefined) {
+                    throw new Error(`${environment.environmentService.getCommandChannelName} command channel not initialized!`);
+                }
                 this.log.info(`stopped environment ${environment.id}.`);
             }
         }
 
         this.commandEmitter.off("command", this.handleCommand);
-        await this.commandChannel.stop();
+        for(const commandChannel of this.commandChannelDict.values()) {
+            commandChannel.stop();
+        }
     }
 
     private async environmentMaintenanceLoop(): Promise<void> {
-        if (this.commandChannel === undefined) {
-            throw new Error(`TrialDispatcher: commandChannel shouldn't be undefined in environmentMaintenanceLoop.`);
-        }
-        const environmentService = component.get<EnvironmentService>(EnvironmentService);
         while (!this.stopping) {
             const environments: EnvironmentInformation[] = [];
             for (const environment of this.environments.values()) {
                 if (environment.isAlive === true) {
                     environments.push(environment);
                 } else {
-                    await this.commandChannel.close(environment);
+                    if (environment.environmentService === undefined) {
+                        throw new Error(`${environment.id} do not have environment service!`);
+                    }
+                    const commandChannel: CommandChannel | undefined = 
+                        this.commandChannelDict.get(environment.environmentService.getCommandChannelName);
+                    if (commandChannel === undefined) {
+                        throw new Error(`${environment.environmentService.getCommandChannelName} command channel not initialized!`);
+                    }
+                    await commandChannel.close(environment);
                 }
             }
-            await environmentService.refreshEnvironmentsStatus(environments);
-
-            environments.forEach((environment) => {
+           
+            for (const environment of environments) {
+                if (environment.environmentService === undefined) {
```

Comment:
Here can group envs by services, and perform one refresh call.

Reply:
updated, group environments first, then update different environments in different environmentService.